tests/drivers/tee/optee/src/main.c

```diff
@@ -0,0 +1,311 @@
+/*
+ * Copyright (c) 2023 Epam Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @addtogroup t_optee_driver
+ * @{
+ * @defgroup t_optee_driver test_optee_driver
+ * @}
+ */
+
+#include <zephyr/arch/arm64/arm-smccc.h>
+#include <zephyr/kernel.h>
+#include <zephyr/ztest.h>
+#include <zephyr/device.h>
+
+#include <zephyr/drivers/tee.h>
+#include "../../../drivers/tee/optee/optee_msg.h"
+#include "../../../drivers/tee/optee/optee_smc.h"
+
+#define TEE_OPTEE_CAP_TZ BIT(0)
+
+typedef void (*smc_cb_t)(unsigned long a0, unsigned long a1, unsigned long a2, unsigned long a3,
+			 unsigned long a4, unsigned long a5, unsigned long a6, unsigned long a7,
+			 struct arm_smccc_res *res);
+
+static struct test_call {
+	int num;
+	smc_cb_t smc_cb;
+	uint32_t a0;
+	uint32_t a1;
+	uint32_t a2;
+	uint32_t a3;
+	uint32_t a4;
+	uint32_t a5;
+	uint32_t a6;
+	uint32_t a7;
+} t_call;
+
+void arm_smccc_smc(unsigned long a0, unsigned long a1, unsigned long a2, unsigned long a3,
+		   unsigned long a4, unsigned long a5, unsigned long a6, unsigned long a7,
+		   struct arm_smccc_res *res)
+{
+	if (t_call.smc_cb) {
+		t_call.smc_cb(a0, a1, a2, a3, a4, a5, a6, a7, res);
+	}
+}
+
+ZTEST(optee_test_suite, test_get_version)
+{
+	int ret;
+	struct tee_version_info info;
+	const struct device *const dev = DEVICE_DT_GET_ONE(linaro_optee_tz);
+
+	zassert_not_null(dev, "Unable to get dev");
+
+	ret = tee_get_version(dev, NULL);
+	zassert_equal(ret, -EINVAL, "tee_get_version failed with code %d", ret);
+
+	ret = tee_get_version(dev, &info);
+	zassert_ok(ret, "tee_get_version failed with code %d", ret);
+	zassert_equal(info.impl_id, 1, "Wrong impl_id");
+	zassert_equal(info.impl_caps, TEE_OPTEE_CAP_TZ, "Wrong impl_caps");
+	zassert_equal(info.gen_caps, TEE_GEN_CAP_GP | TEE_GEN_CAP_REG_MEM,
+		      "Wrong gen_caps");
+
+	ret = tee_get_version(dev, NULL);
+	zassert_equal(ret, -EINVAL, "tee_get_version failed with code %d", ret);
+}
+
+void fast_call(unsigned long a0, unsigned long a1, unsigned long a2, unsigned long a3,
+	       unsigned long a4, unsigned long a5, unsigned long a6, unsigned long a7,
+	       struct arm_smccc_res *res)
+{
+	t_call.a0 = a0;
+	t_call.a1 = a1;
+	t_call.a2 = a2;
+	t_call.a3 = a3;
+	t_call.a4 = a4;
+	t_call.a5 = a5;
+	t_call.a6 = a6;
+	t_call.a7 = a7;
+
+	res->a0 = OPTEE_SMC_RETURN_OK;
+}
+
+void fail_call(unsigned long a0, unsigned long a1, unsigned long a2, unsigned long a3,
+	       unsigned long a4, unsigned long a5, unsigned long a6, unsigned long a7,
+	       struct arm_smccc_res *res)
+{
+	res->a0 = OPTEE_SMC_RETURN_EBUSY;
+}
+
+ZTEST(optee_test_suite, test_fast_calls)
+{
+	int ret;
+	uint32_t session_id;
+	struct tee_open_session_arg arg = {};
+	struct tee_param param = {};
+	const struct device *const dev = DEVICE_DT_GET_ONE(linaro_optee_tz);
+
+	zassert_not_null(dev, "Unable to get dev");
+
+	t_call.num = 0;
+	t_call.smc_cb = fast_call;
```

Comment:
"Fast call" as opposed to the "Normal/Standard call"? Then this is wrong.

Reply:
It's not opposed. I'm just testing the different behaviors. Why it's wrong?

Reply:
Because "open session" is always a standard call, not fast call.

Reply:
And what will be the difference from the unit tests perspective when i set normal call here? 
Considering that unit tests doesn't have connection to the real OP-TEE os - I will have to make a normal call which will just return status OK. This will make it looks the same as fast_call. Here, in the unit tests the intention is to test functions, how they handle different parameters. I've implemented normal_calls when I was in need to test RPC_FUNCS.